zebra/rt_socket.c

```diff
@@ -278,81 +114,143 @@ static int sin6_masklen(struct in6_addr mask)
 #endif /* SIN6_LEN */
 
 /* Interface between zebra message and rtm message. */
-static int kernel_rtm_ipv6(int cmd, const struct prefix *p,
-			   const struct nexthop_group *ng, uint32_t metric)
+static int kernel_rtm(int cmd, const struct prefix *p,
+		      const struct nexthop_group *ng, uint32_t metric)
+
 {
-	struct sockaddr_in6 *mask;
-	struct sockaddr_in6 sin_dest, sin_mask, sin_gate;
+	union sockunion sin_dest, sin_mask, sin_gate;
 #ifdef __OpenBSD__
 	struct sockaddr_mpls smpls;
 #endif
 	union sockunion *smplsp = NULL;
 	struct nexthop *nexthop;
 	int nexthop_num = 0;
 	ifindex_t ifindex = 0;
-	int gate = 0;
+	bool gate = false;
 	int error;
+	char prefix_buf[PREFIX_STRLEN];
 	enum blackhole_type bh_type = BLACKHOLE_UNSPEC;
 
-	memset(&sin_dest, 0, sizeof(struct sockaddr_in6));
-	sin_dest.sin6_family = AF_INET6;
-#ifdef SIN6_LEN
-	sin_dest.sin6_len = sizeof(struct sockaddr_in6);
-#endif /* SIN6_LEN */
-	sin_dest.sin6_addr = p->u.prefix6;
+	if (IS_ZEBRA_DEBUG_RIB)
+		prefix2str(p, prefix_buf, sizeof(prefix_buf));
 
-	memset(&sin_mask, 0, sizeof(struct sockaddr_in6));
+	/*
+	 * We only have the ability to ADD or DELETE at this point
+	 * in time.
+	 */
+	if (cmd != RTM_ADD && cmd != RTM_DELETE) {
+		if (IS_ZEBRA_DEBUG_KERNEL)
+			zlog_debug("%s: %s odd command %s for flags %d",
+				   __func__, prefix_buf,
+				   lookup_msg(rtm_type_str, cmd, NULL),
+				   nexthop->flags);
+		return 0;
+	}
 
-	memset(&sin_gate, 0, sizeof(struct sockaddr_in6));
-	sin_gate.sin6_family = AF_INET6;
+	memset(&sin_dest, 0, sizeof(sin_dest));
+	memset(&sin_gate, 0, sizeof(sin_gate));
+	memset(&sin_mask, 0, sizeof(sin_mask));
+
+	switch (p->family) {
+	case AF_INET:
+		sin_dest.sin.sin_family = AF_INET;
+#ifdef HAVE_STRUCT_SOCKADDR_IN_SIN_LEN
+		sin_dest.sin.sin_len = sizeof(sin_dest);
+		sin_gate.sin.sin_len = sizeof(sin_gate);
+#endif /* HAVE_STRUCT_SOCKADDR_IN_SIN_LEN */
+		sin_dest.sin.sin_addr = p->u.prefix4;
+		sin_gate.sin.sin_family = AF_INET;
+		break;
+	case AF_INET6:
+		sin_dest.sin6.sin6_family = AF_INET6;
+#ifdef SIN6_LEN
+		sin_dest.sin6.sin6_len = sizeof(sin_dest);
+#endif /* SIN6_LEN */
+		sin_dest.sin6.sin6_addr = p->u.prefix6;
+		sin_gate.sin6.sin6_family = AF_INET6;
 #ifdef HAVE_STRUCT_SOCKADDR_IN_SIN_LEN
-	sin_gate.sin6_len = sizeof(struct sockaddr_in6);
+		sin_gate.sin6.sin6_len = sizeof(sin_gate);
 #endif /* HAVE_STRUCT_SOCKADDR_IN_SIN_LEN */
+		break;
+	}
 
 	/* Make gateway. */
 	for (ALL_NEXTHOPS_PTR(ng, nexthop)) {
```

Comment:
Could you reset smplsp to NULL here at the start of the loop? It looks like if it's ever set below, it'll stay set for all subsequent nexthops.

Reply:
yeah fixed.  For future reference I do not believe that ecmp works on most openbsd and mpls labels only work on openbsd.  So this might not be much of a concern but will ensure that it is automatically right for future upgrades to *bsd